mcproto/types/slot.py

```diff
@@ -0,0 +1,80 @@
+from __future__ import annotations
+
+from typing import cast, final
+
+from typing_extensions import override
+
+from mcproto.buffer import Buffer
+from mcproto.protocol import StructFormat
+from mcproto.types.nbt import CompoundNBT, EndNBT, NBTag
+from mcproto.types.abc import MCType
+from attrs import define
+
+
+__all__ = ["Slot"]
+
+"""
+https://wiki.vg/Slot_Data
+"""
+
+
+@define
+@final
+class Slot(MCType):
+    """Represents a slot in an inventory.
+
+    :param present: Whether the slot has an item in it.
+    :param item_id: (optional) The item ID of the item in the slot.
+    :param item_count: (optional) The count of items in the slot.
+    :param nbt: (optional) The NBT data of the item in the slot.
+
+    The optional parameters are present if and only if the slot is present.
+    """
+
+    present: bool
+    item_id: int | None = None
+    item_count: int | None = None
+    nbt: NBTag | None = None
```

Comment:
> I don't really know how to handle 2 or more optional values, and for example here the nbt tag might be None even if present is false.

Well, in this case it's simple, it would be optional in the nested type (namedtuple), but it could be a bit more difficult if there were multiple bool flags for presence indications in other types/packets.

I'm not sure how common something like this is in the rest of the code-base / things that aren't yet done. I know there's a fair amount of optional types here, but are there really that many bool flags that affect multiple variables in the class like this? Cause whenever it just affects a single variable, it's not worth keeping this flag at all, it can be represented directly in the field (by making it optional, or just a union of 2 types, which can be checked for with `isinstance`).

The only time where this becomes  real problem is with these flags affecting multiple variables, which is the case here. If it's just a single such flag, this nested strucutre solution will work well. The issues only arise if there's multiple flags that affect multiple variables which intersect. If they control their own set of multiple variables which don't intersect, it is still easy to solve with nested structures.

With intersections though, it could even mean dozens of such structures as this grows with the cartesian product of those flags, which obviously wouldn't be viable. However, I just really doubt that there actually are many types/packets like this (I don't know of any). But I might be wrong.

> What I think for now is that it's better to not nest the types when they are not nested in the documentation (wiki.vg), so that anyone can just look at wiki.vg's description of the packets/types, and code with them intuitively without having to read our documentation too much.
>
> For this particular case I think it's best to keep the present variable for more clarity and the other next to it directly, but I also understand that it can get become annoying to cast everything.

To be fair, properties would allow that.

> What I suggest is to add a data property that returns the 3 items in a tuple if they are present (and raises an exception otherwise) that way the data can be unpacked without needing typecasts.

hmm, I'm not sure I love that, most people would very likely still use the attributes normally, and this would bring in a lot of overhead if we used it internally (though also if others did) as it'd mean an instantiation each time the function is called, as opposed to if this structure was just stored in the class directly.

> Another option I see is to only guaranty the validity of the data if the present flag is set, by having item_id and item_count set to -1 when present is false (I don't think that would be a good idea for the packets but for the types it seems fine to me). This has the inconvenient of being more error-prone because forgetting to check will lead to silent bugs.

Yeah, I don't like this either, magic values are generally just not a good idea.

> The last option (that I see) would be to have the optional fields hidden behind a property which raises an exception to prevent getting/setting anything while the present field is false. Again I don't think it's a good idea for packets, and this will prevent us from using @define on these classes, while also taking much longer to implement as we would need to define the field, the getter, the setter and the tests for each thing (which can't be tested automatically using gen_serializable_tests).

I don't think this necessarily prevents us from using `@define` it's just that it'd take that inner structure, the only issue is that people would need to manually instantiate that structure first and pass in it, which yeah, could be a bit more annoying than doing so manually, but I'm not sure if that's such a huge issue. The alternative is another constructor or indeed using `__init__` and avoiding `@define`. Depends on whether or not we mind that people will now need to instantiate those inner structures themselves.

I'm not sure why this wouldn't work with tests though, don't they just use `__eq__`? We could just compare against the entire structure, rather than the individual fields. One annoying thing is that it will be slightly less clear as to which field was the invalid one, as the assertion will just show the whole structure now, but the testgen should still work.

The nested structures would just be a lot more idiomatic solution type-wise and the unions here make me kind of uneasy, as they allow for an inconsistent state of the packet. Since the values are only getting validated on runtime after initialization/deserialization, it's entirely possible for anyone to just manually change the `item_id` to `None` after initialization without breaking typing rules, which is worrying.

That said, it's very true that it will take some extra code & time to do and will probably be pretty annoying because of the repetition in the field names in that structure and in the extra properties. So it's just a question whether it's worth that extra work and how it will affect users when working with the packets.

> PS : I also thought about having a master _Slot, a subclass for EmptySlot and one for Slot but this is a no-go (imo) because this would triple the amount of code needed for an otherwise simple class, and make it confusing when using the class anywhere else.

Yeah, this just seems way more convoluted then necessary, we shouldn't make a new class just to represent the potential lack of data, but the thought did cross my mind too.

Python's typing can just be so annoying sometimes, with rust, this could actually be solved very cleanly with the [state pattern](https://refactoring.guru/design-patterns/state), which in rust you can just do with a simple generic attribute, for which you can then have different impls, being completely type safe. (There's a really nice video on this in rust:  <https://www.youtube.com/watch?v=_ccDqRTx-JU> if you're interested, though you might need some basic rust knowledge first.) In OOP langs, you're supposed to just use blank subclasses or composition, which well, isn't ideal.

(Hope you didn't die of old age while reading all this lmao)